core/vm/contracts.go

```diff
@@ -474,88 +432,113 @@ func modExpMultComplexityEip2565(x *big.Int) *big.Int {
 // return multiplication_complexity
 //
 // where is x is max(base_length, modulus_length)
-func modExpMultComplexityEip7883(x *big.Int) *big.Int {
-	if x.Cmp(big32) > 0 {
-		x = modExpMultComplexityEip2565(x)
-		return x.Lsh(x, 1) // ×2
+func modExpMultComplexityEip7883(x uint32) uint64 {
+	if x > 32 {
+		return modExpMultComplexityEip2565(x) * 2
+	}
+	return 16
+}
+
+// modExpMultComplexityEip198 implements modExp multiplication complexity formula, as defined in EIP-198
+//
+// def mult_complexity(x):
+//
+//	if x <= 64: return x ** 2
+//	elif x <= 1024: return x ** 2 // 4 + 96 * x - 3072
+//	else: return x ** 2 // 16 + 480 * x - 199680
+//
+// where is x is max(base_length, modulus_length)
+func modExpMultComplexityEip198(x uint32) uint64 {
+	xx := uint64(x) * uint64(x)
+	switch {
+	case x <= 64:
+		return xx
+	case x <= 1024:
+		// (x ** 2 // 4 ) + ( 96 * x - 3072)
+		return xx/4 + 96*uint64(x) - 3072
+	default:
+		// (x ** 2 // 16) + (480 * x - 199680)
+		// max value: 0x100001df'dffcf220
+		return xx/16 + 480*uint64(x) - 199680
 	}
-	return x.SetUint64(16)
 }
 
 // RequiredGas returns the gas required to execute the pre-compiled contract.
 func (c *bigModExp) RequiredGas(input []byte) uint64 {
+
+	var minGas uint64
+	var adjExpFactor uint64
+	var finalDivisor uint64
+	var calcMultComplexity func(uint32) uint64
+	switch {
+	case c.osaka:
+		minGas = 500
+		adjExpFactor = 16
+		finalDivisor = 1
+		calcMultComplexity = modExpMultComplexityEip7883
+	case c.eip2565:
+		minGas = 200
+		adjExpFactor = 8
+		finalDivisor = 3
+		calcMultComplexity = modExpMultComplexityEip2565
+	default:
+		minGas = 0
+		adjExpFactor = 8
+		finalDivisor = 20
+		calcMultComplexity = modExpMultComplexityEip198
+	}
+
+	header := getData(input, 0, 3*32)
+
+	// If any of the lengths is bigger than uint32, immediately fail.
+	if !allZero(header[0:28]) || !allZero(header[32:32+28]) || !allZero(header[64:64+28]) {
+		// Except for the case where both base and mod are zeros.
+		if allZero(header[0:32]) && allZero(header[64:96]) {
+			return minGas
```

Comment:
This isn't right for Fusaka because `MultComplexity` won't be 0 even when both base and mod are zeros.

Reply:
I'm on it.

Reply:
The issue is hidden (i.e. not testable) because of the EIP-7823 (modexp length limits) applied later in the `Run()` function. In other words: we compute incorrect (by EIP-7883) gas cost here, but it does not matter because the execution ends with an error (by EIP-7823).

This is an argument to move EIP-7823 implementation from `Run()` to `RequiredGas()`.

For now, I'm correcting the implementation.